test/schedulers/custom.ts

```diff
@@ -0,0 +1,286 @@
+      yield * this.#pendingWorkers;
+      yield * this.#readyWorkers;
+    }
+
+    // @ts-expect-error
+    get size () {
+      return this.#pendingWorkers.size + this.#readyWorkers.size;
+    }
+
+    set size (_value) {
+      // no-op
+    }
+
+    onWorkerAvailable (item: PiscinaWorker) {
+      /* istanbul ignore else */
+      if (item.currentUsage() < this.#maxConcurrent) {
+        for (const listener of this.#onAvailableListeners) {
+          listener(item);
+        }
+      }
+    }
+
+    onAvailable (fn: (item: PiscinaWorker) => void) {
+      this.#onAvailableListeners.push(fn);
+    }
+
+    getAvailableCapacity (): number {
+      return this.#pendingWorkers.size * this.#maxConcurrent;
+    }
+  }
+  const maxConcurrent = 1;
+  const pool = new Piscina({
+    filename: resolve(__dirname, '../fixtures/threadid.js'),
+    scheduler: new CustomTaskScheduler(maxConcurrent),
+    maxThreads: 3,
+    minThreads: 3,
+    concurrentTasksPerWorker: maxConcurrent
+  });
+
+  const results = await Promise.all([
+    pool.run(0),
+    pool.run(1),
+    pool.run(2),
+    pool.run(3),
+    pool.run(4),
+    pool.run(5)
+  ]);
+  strictSame(results, [
+    {
+      input: 0,
+      threadId: '1'
+    },
+    {
+      input: 1,
+      threadId: '2'
+    },
+    {
+      input: 2,
+      threadId: '3'
+    },
+    {
+      input: 3,
+      threadId: '1'
+    },
+    {
+      input: 4,
+      threadId: '2'
+    },
+    {
+      input: 5,
+      threadId: '3'
+    }
+  ]);
+});
+
+test('Should throw if bad CustomTaskScheduler', ({ plan, throws }) => {
+  plan(1);
+  const maxConcurrent = 1;
+  throws(
+    () =>
+      new Piscina({
+        filename: resolve(__dirname, '../fixtures/threadid.js'),
+        // @ts-expect-error
+        scheduler: {},
+        maxThreads: 3,
+        minThreads: 3,
+        concurrentTasksPerWorker: maxConcurrent
+      }),
+    TypeError,
+    'options.scheduler must be a valid scheduler instance'
+  );
+});
+
+test('Should accept CustomTaskSchedulers that does not inherit from base', async ({
+  plan,
+  strictSame
+}) => {
+  const maxConcurrent = 1;
+  // Simplistic RoundRobin example
+  class CustomTaskScheduler {
+    #maxConcurrent: number;
+    #readyWorkers: Set<PiscinaWorker>;
+    #pendingWorkers: Set<PiscinaWorker>;
+    // TODO: create its own type?
+    #onAvailableListeners: ((item: PiscinaWorker) => void)[];
+    #offset: number;
+
+    constructor (maxConcurrent: number) {
+      this.#maxConcurrent = maxConcurrent;
+      this.#readyWorkers = new Set();
+      this.#pendingWorkers = new Set();
+      this.#onAvailableListeners = [];
+      this.#offset = 0;
+    }
+
+    add (item) {
+      this.#pendingWorkers.add(item);
+      item.onReady(() => {
+        /* istanbul ignore else */
+        if (this.#pendingWorkers.has(item)) {
+          this.#pendingWorkers.delete(item);
+          this.#readyWorkers.add(item);
+          this.onWorkerAvailable(item);
+        }
```

Comment:
```suggestion
        if (this.#pendingWorkers.delete(item)) {
          this.#readyWorkers.add(item);
          this.onWorkerAvailable(item);
        }
```

Reply:
Done in [0b58d1c](https://github.com/piscinajs/piscina/pull/487/commits/0b58d1c2633033ece1bbcb40bacecfc2c40d63c3)